File: ios/stereoCamera/stereoCamera/Comm/Commands/SetCaptureQuality.swift
//
//  SetCaptureQuality.swift
//  stereoCamera
//
//  Created by hallmarklabs on 8/1/18.
//  Copyright © 2018 cody. All rights reserved.
//

import Foundation

class SetCaptureQuality : Command
{
    var quality:ImageQuality = ImageQuality.LOW
    
    init(_ quality:ImageQuality)
    {
        super.init()
        
        self.quality = quality
        doInit()
    }
    
    override init()
    {
        super.init()
        doInit()
    }
    
    func doInit()
    {
        cmdtype = CommandTypes.SET_CAPTURE_QUALITY
        expectsResponse = true
    }
    
    override func send(comm: Comm) -> Bool
    {
        let result = super.send(comm: comm)
        if (!result) { return false }
        
        var bytes:[UInt8] = []
        bytes += Bytes.toByteArray(quality)

        let sz = comm.write(buf: bytes)
        return (sz <= 0) ? false : true
    }
    
    override func receive(comm: Comm) -> Bool
    {
        let result = super.receive(comm: comm)
        if (!result) { return false }
       
        let bytes = [UInt8].init(repeating: 0, count: 1)
        let sz = comm.read(buffer: bytes)
        if (sz <= 0) { return false }
        
        quality = ImageQuality(rawValue: Int(bytes[0]))!
        return true
    }
}

File: android/app/src/main/java/com/munger/stereocamera/ip/command/commands/SetCaptureQuality.java
package com.munger.stereocamera.ip.command.commands;

import com.munger.stereocamera.ip.command.Comm;
import com.munger.stereocamera.ip.command.Command;
import com.munger.stereocamera.widget.PreviewWidget;

import java.io.IOException;

public class SetCaptureQuality extends Command
{
	public PreviewWidget.ShutterType quality;

	public SetCaptureQuality()
	{
		super();
		doInit();
	}

	public SetCaptureQuality(PreviewWidget.ShutterType quality)
	{
		super();
		this.quality = quality;
		doInit();
	}

	private void doInit()
	{
		cmdtype = Type.SET_CAPTURE_QUALITY;
		expectsResponse = true;
	}

	@Override
	public boolean receive(Comm comm)
	{
		boolean result = super.receive(comm);
		if (!result)
			{ return false; }

		try
		{
			byte b = comm.getByte();
			PreviewWidget.ShutterType[] quals = PreviewWidget.ShutterType.values();

			if (b < 0 || b >= quals.length)
				{ return false; }

			quality = quals[b];
		}
		catch(IOException e ) {
			return false;
		}

		return true;
	}

	@Override
	public boolean send(Comm comm)
	{
		boolean result = super.send(comm);
		if (!result)
			{ return false; }

		try
		{
			byte val = (byte) quality.ordinal();
			comm.putByte(val);
		}
		catch(IOException e){
			return false;
		}

		return true;
	}
}
